src/mrack/providers/provider.py

```diff
@@ -408,23 +435,58 @@ async def strategy_retry(self, reqs):
                 break
 
             attempts += 1
+
+            # set the waiting timeout to vary from 45-75 minutes randomly
+            # so if there are multiple parallel runs of mrack they wait
+            # different amount time before poll and increase chance to get resources
+            # quicker than other concurrent runs of mrack requests
+            delta = random.choice(range(-delta_sleep, delta_sleep))
+            res_check_timeout = self.timeout + delta
+
             s_hosts, error_hosts, missing_reqs = await self._provision_base(
-                missing_reqs
+                missing_reqs, timeout=res_check_timeout
             )
             success_hosts.extend(s_hosts)
 
             # We need to cleanup the error hosts for the next retry run
             # in case of last attempt which is `self.max_retry`
             # we skip this part at it done in provision_hosts() method while
             # awaiting the self.abort_and_delete(error hosts)
+
             if error_hosts and attempts <= self.max_retry:
                 count = len(error_hosts)
                 err = f"{count} hosts were not provisioned properly, deleting."
                 logger.info(f"{log_msg_start} {err}")
+                out_of_resources = False
                 for host in error_hosts:
                     logger.error(f"{log_msg_start} Error: {str(host.error)}")
+                    if host.error.get("code", None) == 500:  # 500 = SERVER ERROR
+                        logger.info(
+                            f"{log_msg_start} Provider is out of resources, "
+                            "increasing cooldown time before another retry"
+                        )
+                        out_of_resources = True
+
+                if await self.utilization() >= max_utilization or out_of_resources:
+                    # delete all hosts when there is high utilization of provider
+                    error_hosts += success_hosts
```

Comment:
Just wondering: Wouldn't it be better to check in addition:

if percentage of hosts already provisioned is low, then free resources. If our host list is close to be completely provisioned then we would not free resources.

This way we would get optimized result across concurrent runs?